File: apps/java/io/dustin/apps/board/api/usecase/community/posting/ReadPostingUseCase.java
package io.dustin.apps.board.api.usecase.community.posting;

import io.dustin.apps.board.api.usecase.community.comment.ReadCommentUseCase;
import io.dustin.apps.board.domain.community.comment.model.dto.CommentDto;
import io.dustin.apps.board.domain.community.comment.service.ReadCommentService;
import io.dustin.apps.board.domain.community.posting.model.dto.PostingDetailDto;
import io.dustin.apps.board.domain.community.posting.model.dto.PostingDto;
import io.dustin.apps.board.domain.community.posting.model.dto.PostingListDto;
import io.dustin.apps.board.domain.community.posting.service.ReadPostingService;
import io.dustin.apps.board.domain.community.posting.service.WritePostingService;
import io.dustin.apps.common.model.CountByPagingInfo;
import io.dustin.apps.common.model.QueryPage;
import io.dustin.apps.common.model.ResponseWithScroll;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

import static io.dustin.apps.common.model.ResponseWithScrollSetting.getCountByPagingInfo;

@Service
@RequiredArgsConstructor
public class ReadPostingUseCase {

    private final ReadPostingService readPostingService;
    private final ReadCommentService readCommentService;
    private final WritePostingService writePostingService;
    private final ReadCommentUseCase readCommentUseCase;

    public ResponseWithScroll<List<PostingListDto>> execute(QueryPage queryPage) {
        /**
         * 게시물 리스트 보여줌
         */
        long userId = 1;
        int realSize = queryPage.getSize();
        int querySize = realSize + 1;

        List<PostingListDto> result = readPostingService.getPostingList(userId, queryPage.getNextId(), querySize);
        CountByPagingInfo<PostingListDto> cbi = getCountByPagingInfo(result, realSize);

        return ResponseWithScroll.from(cbi.result(), cbi.isLast(), cbi.nextId());

    }

    @Transactional
    public PostingDetailDto postingDetail(Long postingId, QueryPage queryPage) {
        /**
         * 게시물 상세조회 했을때 게시물과 댓글을 같이 보여줌
         */
        long userId = 1;

        writePostingService.click(postingId);
        PostingDto posting = readPostingService.getPosting(userId, postingId);
        ResponseWithScroll<List<CommentDto>> commentList = readCommentUseCase.execute(postingId,queryPage);

        return PostingDetailDto.from(posting,commentList);



        /** 추후 재사용 */
//        int realSize = queryPage.getSize();
//        int querySize = realSize + 1;
//
//        List<CommentDto> result = readCommentService.getCommentsByPosting(userId, postingId, querySize, queryPage.getNextId());
//        CountByPagingInfo<CommentDto> cbi = getCountByPagingInfo(result, realSize);
//
//        if(result.size() == 1) {
//            Long commentId = result.get(0).id();
//            if(commentId == null) {
//                return PostingDetailDtoTest.from(result.get(0).posting(), ResponseWithScroll.from(Collections.emptyList(), cbi.isLast(), cbi.nextId()));
//            }
//        }
//
//        ResponseWithScroll<List<CommentDto>> commentList =  ResponseWithScroll.from(cbi.result(), cbi.isLast(), cbi.nextId());
//        Map<Posting, List<CommentDto>> map = cbi.result().stream()
//                                                         .collect(groupingBy(CommentDto::posting));
//        return map.entrySet().stream()
//                             .map(Map.Entry::getKey)
//                             .map( posting -> PostingDetailDtoTest.from(posting, commentList))
//                             .findFirst().orElseThrow();
    }

}

File: apps/src/main/kotlin/io/dustin/apps/board/api/usecase/community/posting/ReadPostingUseCase.kt
package io.dustin.apps.board.api.usecase.community.posting

import io.dustin.apps.board.api.community.request.query.AllPostingQuery
import io.dustin.apps.board.api.community.request.query.PostingDetailQuery
import io.dustin.apps.board.api.usecase.community.comment.ReadCommentUseCase
import io.dustin.apps.board.domain.community.comment.model.dto.CommentDto
import io.dustin.apps.board.domain.community.posting.model.dto.PostingDetailDto
import io.dustin.apps.board.domain.community.posting.model.dto.PostingDto
import io.dustin.apps.board.domain.community.posting.model.dto.PostingListDto
import io.dustin.apps.board.domain.community.posting.service.ReadPostingService
import io.dustin.apps.board.domain.community.posting.service.WritePostingService
import io.dustin.apps.common.model.QueryPage
import io.dustin.apps.common.model.ResponseWithScroll
import io.dustin.apps.common.model.response.ResultResponse
import io.dustin.apps.common.model.response.ResultResponsePagination
import io.dustin.apps.common.utils.lastAt
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class ReadPostingUseCase (
    private val readPostingService: ReadPostingService,
    private val writePostingService: WritePostingService,
    private val readCommentUseCase: ReadCommentUseCase
) {
    @Transactional
    fun allPostingList(query: AllPostingQuery): ResultResponsePagination<PostingListDto> {
        val userId: Long = 1
        val list = with(query){
            readPostingService.getPostingList(userId, query.recordsCount, query.nextId)
        }
        println("몇개나 들어있을까요" + list.size)
        if(list.isEmpty()) {
            return ResultResponsePagination.of(
                last = true,
                recordsCount = 0,
                data = list,
            )
        }
        val result = list.take((query.recordsCount).toInt())
        val isLast = list.size <= query.recordsCount.toInt()

        val nextId = if (!isLast) {
            result.lastAt().id
        } else {
            null
        }

        return ResultResponsePagination.of(
            nextId = nextId,
            last = isLast,
            recordsCount = result.size.toLong(),
            data = result,
        )


    }
    @Transactional
    fun postingDetail(query: PostingDetailQuery): ResultResponse<PostingDetailDto> {

        val userId: Long = 1
        writePostingService.click(query.postingId)
        val posting: PostingDto = readPostingService.getPosting(userId, query.postingId)
        val commentList= readCommentUseCase.execute(query)
        val result = PostingDetailDto.from(posting, commentList)
        return ResultResponse.of(result)
    }

}

